### PR TITLE
Integrate cert management for Shoot clusters

### LINKED ISSUE #449: Cert-Broker integration
Now that we have open-sourced the [cert-broker](https://github.com/gardener/cert-broker) we need to integrate it into Gardener.

Gardener needs to deploy the [cert-manager](https://github.com/jetstack/cert-manager) once into the seed cluster's `garden` namespace, and the cert-broker once per shoot control plane.
We will only allow DNS-based certificate challenges.

End-users of shoot clusters can then trigger the cert-broker to issue trusted certificates of Let's Encrypt for their `Ingress`es by annotating them properly (needs to be well-documented). With that, end-users can use the Let's Encrypt account of the Gardener provider instead of bringing their own. However, this functionality should only apply if a default domain is used.

### PR DESCRIPTION
**Pull request by timuthy**
**What this PR does / why we need it**:
This PR integrates the components [Cert-Manager](https://github.com/jetstack/cert-manager) and [Cert-Broker](https://github.com/gardener/cert-broker) which handle certificate management for Shoot clusters. While the Cert-Manager is deployed once per Seed in the `Garden` namespace, the Cert-Broker is part of every Shoot cluster's control plane.

This feature can be enabled by setting the `CertificateManagement: true` in the `featureGates` section of `ControllerManagementConfiguration`. If enabled, the following information must be provided:

- Access information for domains which certificates should be issued for.
- Optional: Access Information for `Let's Encrypt` account to be used.

**Which issue(s) this PR fixes**:
Fixes #449

**Release note**:

```noteworthy user
Gardener can now issue valid certificates for `Ingress`es specifying having a default domains and those labelled with `garden.sapcloud.io/purpose=managed-cert`.
```
```noteworthy operator
Gardener can now manage `Ingress` resources requesting a certificate for Gardener's default domains. This can be enabled by enabling the `CertificateManagement` feature gate in the `ControllerManagementConfiguration`. Please make sure that the required configuration is placed in the `garden` namespace of the garden cluster (please find [here](https://github.com/gardener/gardener/blob/master/example/10-secret-certificate-management-config.yaml) an example configuration).
```

### CODE REVIEW

File: charts/seed-bootstrap/values.yaml [hunk 1]
```diff
@@ -77,3 +78,6 @@ fluentd-es:
   enabled: true
   images:
     fluentd-es: image-repository:image-tag
+
+cert-manager:
+  enabled: true
```
Comment (adracus) [inline charts/seed-bootstrap/values.yaml:85]: how about naming this `certificate-management` like in the cert-broker?

File: pkg/operation/seed/seed.go [hunk 2]
```diff
@@ -298,9 +325,147 @@ func BootstrapCluster(seed *Seed, k8sGardenClient kubernetes.Client, secrets map
...
+		if err != nil {
+			return err
+		}
+		logger.Logger.Debugf("%s secret was created successfully", secret.Name)
+	}
+	return nil
+}
+
+func createDNSProviders(configs []certmanagement.DNSProviderConfig) ([]interface{}, []*corev1.Secret) {
+	var providers []interface{}
```
Comment (adracus) [inline pkg/operation/seed/seed.go:383]: please use a `var` block for this

File: pkg/operation/seed/seed.go [hunk 3]
```diff
@@ -298,9 +325,147 @@ func BootstrapCluster(seed *Seed, k8sGardenClient kubernetes.Client, secrets map
...
+	var secrets []*corev1.Secret
+	for _, config := range configs {
+		name := config.ProviderName()
```
Comment (adracus) [inline pkg/operation/seed/seed.go:385]: also `var` block here for these many variables

File: pkg/operation/seed/seed.go [hunk 4]
```diff
@@ -298,9 +325,147 @@ func BootstrapCluster(seed *Seed, k8sGardenClient kubernetes.Client, secrets map
...
+		secret, err := k8sSeedClient.CreateSecretObject(clusterIssuerSecret, true)
+		if err != nil {
+			return err
+		}
+		logger.Logger.Debugf("%s secret was created successfully", secret.Name)
+	}
+	return nil
+}
+
+func createDNSProviders(configs []certmanagement.DNSProviderConfig) ([]interface{}, []*corev1.Secret) {
```
Comment (adracus) [inline pkg/operation/seed/seed.go]: Split this function into one that generates secrets and one that generates helm configuration, then also the name can improve.

File: pkg/operation/botanist/controlplane.go [hunk 5]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+
+	certBroker, err := b.InjectImages(certBrokerConfig, b.K8sSeedClient.Version(), b.K8sSeedClient.Version(), common.CertBrokerImageName)
+	if err != nil {
+		return nil
+	}
+
+	return b.ApplyChartSeed(filepath.Join(chartPathControlPlane, "cert-broker"), "cert-broker", b.Shoot.SeedNamespace, nil, certBroker)
+}
+
+func createDNSProviders(config certmanagement.DNSProviderConfig, existingDNSConfig *[]interface{}) {
```
Comment (adracus) [inline pkg/operation/botanist/controlplane.go]: don't mutate the `existingDNSConfig` here, rather return a `[]interface` array and append at the place where you're using this function.

**Review by adracus — CHANGES_REQUESTED** [review]
Some code-wise improvements

File: charts/seed-bootstrap/charts/cert-manager/templates/clusterissuer.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    dns01:
+      providers:
+      {{- range .Values.clusterissuer.acme.dns01.providers }}
+      - name: {{ .name }}
+      {{- if eq .type "aws-route53" }}
+        route53:
+          region: {{ .region }}
+          accessKeyID: {{ .accessKeyID }}
+          secretAccessKeySecretRef:
+              name: {{ .secretName }}
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/cert-manager/templates/clusterissuer.yaml]: indentation

File: charts/seed-bootstrap/charts/cert-manager/templates/rbac.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    app: {{ template "cert-manager.name" . }}
+    chart: {{ template "cert-manager.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+rules:
+  - apiGroups: ["certmanager.k8s.io"]
+    resources: ["certificates", "issuers", "clusterissuers"]
+    verbs: ["*"]
+  - apiGroups: [""]
+    # TODO: remove endpoints once 0.4 is released. We include it here in case
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/cert-manager/templates/rbac.yaml]: Can we remove this TODO?

File: charts/seed-bootstrap/requirements.yaml [hunk 8]
```diff
@@ -9,3 +9,8 @@ dependencies:
   version: 0.1.0
   # needed for the logging feature gate
   condition: fluentd-es.enabled
+- name: cert-manager
+  repository: http://localhost:10191
+  version: 0.4.1
+  # needed for the logging feature gate
```
Comment (rfranzke) [inline charts/seed-bootstrap/requirements.yaml]: s/logging/CertificateManagement

File: charts/seed-bootstrap/values.yaml (hunk 1 above)
Comment (rfranzke) [reply charts/seed-bootstrap/values.yaml:85]: At this place it makes sense to name it `cert-manager` (as only the cert-manager is installed), right?

File: charts/seed-controlplane/charts/cert-broker/.helmignore [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Patterns to ignore when building packages.
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/cert-broker/.helmignore]: Remove this file, not needed.

File: charts/shoot-core/charts/cert-broker/templates/rbac.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRoleBinding
+metadata:
+  name: system:cert-broker
```
Comment (rfranzke) [inline charts/shoot-core/charts/cert-broker/templates/rbac.yaml]: Better: `garden.sapcloud.io:system:cert-broker`

File: charts/shoot-core/charts/cert-broker/templates/rbac.yaml [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  kind: ClusterRole
+  name: system:cert-broker
+subjects:
+- kind: User
+  name: system:cert-broker
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRole
+metadata:
+  name: system:cert-broker
```
Comment (rfranzke) [inline charts/shoot-core/charts/cert-broker/templates/rbac.yaml]: Better: `garden.sapcloud.io:system:cert-broker`

File: example/25-secret-certificate-management-config.yaml [hunk 12]
```diff
@@ -0,0 +1,44 @@
+---
```
Comment (rfranzke) [inline example/25-secret-certificate-management-config.yaml]: Rename to `10-secret-...` (all secrets of this kind have the `10` prefix).

File: example/25-secret-certificate-management-config.yaml [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+kind: Secret
+metadata:
+  name: certificate-management
+  namespace: garden
+  labels:
+    garden.sapcloud.io/role: certificate-management
+type: Opaque
+data:
+  certificate-management-config: base64(certificate-management-config)
+# ---
```
Comment (rfranzke) [inline example/25-secret-certificate-management-config.yaml]: Add a newline before and quickly explain that the following is an example for a certificate management config.

File: pkg/client/kubernetes/base/rbac.go [hunk 14]
```diff
@@ -58,3 +58,13 @@ func (c *Client) patchRoleBinding(oldObj, newObj *rbacv1.RoleBinding) (*rbacv1.R
 
 	return c.clientset.RbacV1().RoleBindings(oldObj.Namespace).Patch(oldObj.Name, types.StrategicMergePatchType, patch)
 }
+
+// DeleteClusterRole deltes a ClusterRole object.
```
Comment (rfranzke) [inline pkg/client/kubernetes/base/rbac.go]: typo

File: pkg/client/kubernetes/base/rbac.go [hunk 15]
```diff
@@ -58,3 +58,13 @@ func (c *Client) patchRoleBinding(oldObj, newObj *rbacv1.RoleBinding) (*rbacv1.R
...
+func (c *Client) DeleteClusterRole(name string) error {
+	return c.Clientset().RbacV1().ClusterRoles().Delete(name, &defaultDeleteOptions)
+}
+
+// DeleteRoleBinding deltes a RoleBindung object.
```
Comment (rfranzke) [inline pkg/client/kubernetes/base/rbac.go]: typo

File: pkg/controller/shoot/shoot.go [hunk 16]
```diff
@@ -117,7 +115,7 @@ func NewShootController(k8sGardenClient kubernetes.Client, k8sGardenInformers ga
 		config:             config,
 		control:            NewDefaultControl(k8sGardenClient, gardenv1beta1Informer, secrets, imageVector, identity, config, gardenNamespace, recorder),
 		careControl:        NewDefaultCareControl(k8sGardenClient, gardenv1beta1Informer, secrets, imageVector, identity, config),
-		maintenanceControl: NewDefaultMaintenanceControl(k8sGardenClient, gardenv1beta1Informer, secrets, imageVector, identity, recorder),
+		maintenanceControl: NewDefaultMaintenanceControl(k8sGardenClient, gardenv1beta1Informer, secrets, imageVector, identity, recorder, config),
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot.go]: Why is this added? The maintenance controller does not need it.

File: pkg/controller/shoot/shoot_maintenance_control.go [hunk 17]
```diff
@@ -135,8 +135,8 @@ type MaintenanceControlInterface interface {
 // implements the documented semantics for maintaining Shoots. updater is the UpdaterInterface used
 // to update the spec of Shoots. You should use an instance returned from NewDefaultMaintenanceControl() for any
 // scenario other than testing.
-func NewDefaultMaintenanceControl(k8sGardenClient kubernetes.Client, k8sGardenInformers gardeninformers.Interface, secrets map[string]*corev1.Secret, imageVector imagevector.ImageVector, identity *gardenv1beta1.Gardener, recorder record.EventRecorder) MaintenanceControlInterface {
-	return &defaultMaintenanceControl{k8sGardenClient, k8sGardenInformers, secrets, imageVector, identity, recorder}
+func NewDefaultMaintenanceControl(k8sGardenClient kubernetes.Client, k8sGardenInformers gardeninformers.Interface, secrets map[string]*corev1.Secret, imageVector imagevector.ImageVector, identity *gardenv1beta1.Gardener, recorder record.EventRecorder, config *componentconfig.ControllerManagerConfiguration) MaintenanceControlInterface {
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_maintenance_control.go]: Why is this added? The maintenance controller does not need it.

File: pkg/controller/shoot/shoot_control_reconcile.go [hunk 18]
```diff
@@ -210,7 +210,16 @@ func (c *defaultControl) reconcileShoot(o *operation.Operation, operationType ga
 			Fn:           flow.SimpleTaskFn(botanist.DeployClusterAutoscaler).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(reconcileMachines, deployKubeAddonManager, deploySeedMonitoring),
 		})
-
+		_ = g.Add(flow.Task{
+			Name:         "Deploying Cert-Broker",
+			Fn:           flow.SimpleTaskFn(botanist.DeployCertBroker).DoIf(certManagementEnabled).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(initializeShootClients, deployKubeAddonManager),
+		})
+		_ = g.Add(flow.Task{
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control_reconcile.go]: Please remove this task and add the logic to the `DeployCertBroker` function (like [here](https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/controlplane.go#L195-L197)).

File: pkg/operation/botanist/controlplane.go [hunk 19]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
 	}
 	return nil
 }
+
+// DeployCertBroker deploys the Cert-Broker to the Shoot namespace in the Seed.
+func (b *Botanist) DeployCertBroker() error {
+	certificateManagement, ok := b.Operation.Secrets[common.GardenRoleCertificateManagement]
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: s/`b.Operation.Secrets`/`b.Secrets`

File: pkg/operation/botanist/controlplane.go [hunk 20]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+		*existingDNSConfig = append(*existingDNSConfig, map[string]interface{}{
+			"domain":   domain,
+			"provider": config.ProviderName(),
+		})
+	}
+}
+
+// DeleteCertBroker delete the Cert-Broker deployment if cert-management in disabled.
+func (b *Botanist) DeleteCertBroker() error {
+	logger.Logger.Info("Deleting cert-broker deployment as cert-management is disabled")
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: Remove the logging, we don't have it anywhere else

File: pkg/operation/botanist/controlplane.go [hunk 21]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+	err := b.K8sSeedClient.DeleteDeployment(b.Shoot.SeedNamespace, common.CertBrokerResourceName)
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: One line:

```golang
if err := b.K8sSeedClient.DeleteDeployment(b.Shoot.SeedNamespace, common.CertBrokerResourceName); err != nil && !apierrors.IsNotFound(err) {
  return err
}
```

File: pkg/operation/botanist/controlplane.go [hunk 22]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	err = b.K8sSeedClient.DeleteSecret(b.Shoot.SeedNamespace, common.CertBrokerResourceName)
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: one line

File: pkg/operation/botanist/controlplane.go [hunk 23]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	err = b.K8sSeedClient.DeleteServiceAccount(b.Shoot.SeedNamespace, common.CertBrokerResourceName)
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: one line

File: pkg/operation/botanist/controlplane.go [hunk 24]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	err = b.K8sSeedClient.DeleteRoleBinding(b.Shoot.SeedNamespace, common.CertBrokerResourceName)
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: one line

File: pkg/operation/botanist/controlplane.go [hunk 25]
```diff
@@ -470,3 +474,80 @@ func (b *Botanist) WakeUpControllers() error {
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	err = b.K8sSeedClient.DeleteClusterRole(common.CertBrokerResourceName)
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: one line

File: pkg/operation/botanist/secrets.go [hunk 26]
```diff
@@ -635,6 +635,25 @@ func (b *Botanist) generateWantedSecrets(basicAuthAPIServer *secrets.BasicAuth,
...
 	}
 
+	certManagementEnabled := features.ControllerFeatureGate.Enabled(features.CertificateManagement)
+	if certManagementEnabled {
+		secretList = append(secretList,
+			&secrets.ControlPlaneSecretConfig{
+				CertificateSecretConfig: &secrets.CertificateSecretConfig{
+					Name: common.CertBrokerResourceName,
+
+					CommonName: "system:cert-broker",
```
Comment (rfranzke) [inline pkg/operation/botanist/secrets.go]: `garden.sapcloud.io:system:cert-broker`

File: pkg/operation/common/types.go [hunk 27]
```diff
@@ -72,6 +72,10 @@ const (
 	// DNS Hosted Zone.
 	DNSHostedZoneID = "dns.garden.sapcloud.io/hostedZoneID"
 
+	// DNSGoogleCloudDNSProject is the key for an annotation on a Kubernetes Secret object whose value must point to a valid
+	// Google Cloud DNS project.
+	DNSGoogleCloudDNSProject = "dns.garden.sapcloud.io/cloudDNSProject"
```
Comment (rfranzke) [inline pkg/operation/common/types.go]: add `google` to the name, e.g. `dns.garden.sapcloud.io/google-clouddns-project`

File: pkg/operation/seed/seed.go [hunk 28]
```diff
@@ -261,6 +268,26 @@ func BootstrapCluster(seed *Seed, k8sGardenClient kubernetes.Client, secrets map
...
+		if !ok {
+			return fmt.Errorf("certificate management is enabled but no secret could be found with role: %s", common.GardenRoleCertificateManagement)
+		}
+
+		clusterIssuer, err = createClusterIssuer(k8sSeedClient, certificateManagement)
+		if err != nil {
+			return fmt.Errorf("cannot create Cluster Issuer for certificate management: %v", err)
+		}
+	} else {
+		err := k8sSeedClient.DeleteDeployment(common.GardenNamespace, common.CertManagerResourceName)
```
Comment (rfranzke) [inline pkg/operation/seed/seed.go]: one line

File: pkg/operation/seed/seed.go [hunk 29]
```diff
@@ -298,9 +325,147 @@ func BootstrapCluster(seed *Seed, k8sGardenClient kubernetes.Client, secrets map
...
+				Namespace: common.GardenNamespace,
+			},
+			Data: map[string][]byte{
+				"tls.key": []byte(acmePrivateKey),
+			},
+		}
+		secrets = append(secrets, letsEncryptKeySecret)
+	}
+
+	err = createCertificateIssuerSecrets(k8sSeedclient, secrets)
```
Comment (rfranzke) [inline pkg/operation/seed/seed.go]: one line

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: charts/seed-bootstrap/values.yaml (hunk 1 above)
Comment (timuthy) [reply charts/seed-bootstrap/values.yaml:85]: Agree with @rfranzke and renamed `certificate-management` to `cert-broker` for Shoot core addons to be consistent.

**Comment by timuthy** [comment]
Thanks for your review @Adracus , @rfranzke. I've addressed your comments, PTAL.

File: charts/seed-bootstrap/charts/cert-manager/templates/clusterissuer.yaml [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+    dns01:
+      providers:
+      {{- range .Values.clusterissuer.acme.dns01.providers }}
+      - name: {{ .name }}
+      {{- if eq .type "aws-route53" }}
+        route53:
+          region: {{ .region }}
+          accessKeyID: {{ .accessKeyID }}
+          secretAccessKeySecretRef:
+            name: {{ .secretName }}
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/cert-manager/templates/clusterissuer.yaml]: As discussed: Let's create the secrets here as part of this Helm chart instead of doing it in the code.

File: charts/seed-bootstrap/charts/cert-manager/values.yaml [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+    email: john.doe@example.com
+    server: https://acme-v02.api.letsencrypt.org/directory
+    privateKeySecretRef: issuer
+    dns01:
+      providers:
+      - name: prod-route53
+        type: route53
+        region: us-east-1
+        accessKeyID:
+        secretName:
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/cert-manager/values.yaml]: This should not be required to be provided but can be computed within the Helm chart.

File: charts/seed-bootstrap/charts/cert-manager/values.yaml [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+        secretKey:
+      - name: prod-clouddns
+        type: google-clouddns
+        project:
+        secretName:
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/cert-manager/values.yaml]: This should not be required to be provided but can be computed within the Helm chart.

File: example/10-secret-certificate-management-config.yaml [hunk 33]
```diff
@@ -0,0 +1,45 @@
+---
```
Comment (rfranzke) [inline example/10-secret-certificate-management-config.yaml:1]: As discussed: Please add this to the Gardener Helm chart so that it is not needed to be installed/deployed separately.

File: pkg/operation/seed/seed.go (hunk 28 above)
Comment (rfranzke) [reply pkg/operation/seed/seed.go]: You resolved the comment, but you didn't change it. One line, please. :)

File: pkg/operation/seed/seed.go [hunk 34]
```diff
@@ -298,9 +325,155 @@ func BootstrapCluster(seed *Seed, k8sGardenClient kubernetes.Client, secrets map
...
+				logger.Logger.Errorf("Failed to cast to Route53Config object for DNSProviderConfig  %+v", config)
+				return nil
+			}
+
+			providers = append(providers, map[string]interface{}{
+				"name":        name,
+				"type":        certmanagement.Route53,
+				"region":      route53config.Region,
+				"accessKeyID": route53config.AccessKeyID,
+				"secretName":  name,
```
Comment (rfranzke) [inline pkg/operation/seed/seed.go]: As said above: When you create the secrets via the Helm chart this name is not needed any more (it can be computed within the Helm chart).

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by timuthy** [comment]
@rfranzke, I've addressed your comments, PTAL.

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Comment by rfranzke** [comment]
Please rebase once again.

**Review by adracus — APPROVED** [review]
\lgtm

**Review by rfranzke — APPROVED** [review]
